eip712/src/main/java/pm/gnosis/eip712/EIP712JsonParser.kt

```diff
@@ -0,0 +1,83 @@
+package pm.gnosis.eip712
+
+import pm.gnosis.model.Solidity
+import pm.gnosis.utils.hexAsBigInteger
+import pm.gnosis.utils.hexStringToByteArray
+import java.io.InputStream
+import java.math.BigInteger
+
+data class DomainWithMessage(val signer: Solidity.Address, val domain: Struct712, val message: Struct712)
+
+class EIP712JsonParser(private val jsonAdapter: JsonAdapter) {
+    fun parseMessage(rawJson: String): DomainWithMessage = parseMessage(jsonAdapter.parse(rawJson))
+
+    fun parseMessage(inputStream: InputStream): DomainWithMessage = parseMessage(jsonAdapter.parse(inputStream))
+
+    private fun parseMessage(adapterResult: JsonAdapter.Result): DomainWithMessage {
+        val primaryTypeParams = adapterResult.types[adapterResult.primaryType]
+            ?: throw IllegalArgumentException("TypedData does not contain primary type ${adapterResult.primaryType}")
+
+        val domainTypeParams = adapterResult.types[DOMAIN_TYPE]
+            ?: throw IllegalArgumentException("TypedData does not contain a domain")
+
+        return DomainWithMessage(
+            signer = adapterResult.signer,
+            domain = buildStruct712(
+                type = DOMAIN_TYPE,
+                params = domainTypeParams,
+                values = adapterResult.domain,
+                typeSpec = adapterResult.types
+            ),
+            message = buildStruct712(
+                type = adapterResult.primaryType,
+                params = primaryTypeParams,
+                values = adapterResult.message,
+                typeSpec = adapterResult.types
+            )
+        )
+    }
+
+    private fun buildStruct712(
+        type: String,
+        params: List<JsonAdapter.TypeParam>,
+        values: Map<String, Any>,
+        typeSpec: Map<String, List<JsonAdapter.TypeParam>>
+    ): Struct712 {
+        val innerParams = params.map { typeParam ->
+            typeParam.name to
+                    if (!solidityTypes.contains(typeParam.type)) {
+                        // Struct
+                        buildStruct712(
+                            type = typeParam.type,
+                            params = typeSpec[typeParam.type]!!,
+                            values = values[typeParam.name] as Map<String, Any>,
+                            typeSpec = typeSpec
+                        )
+                    } else {
+                        // Literal
+                        // TODO this doesn't check size constraints ie.: having uint8 with a value greater than 8 bits
+                        val value = values[typeParam.name].toString()
+                        val bivrostType = when {
+                            typeParam.type.startsWith(prefix = "uint") -> Solidity.UInt256(readNaturalNumber(value))
```

Comment:
check number (everthing > 256 should be invalid)

Reply:
Fixed in https://github.com/gnosis/svalinn-kotlin/pull/31/commits/26eb7880abd8216cb6d2cfd20ce7a9cdad1441ab